# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [30]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [4]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [7]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [13]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [14]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [4]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score.

In [16]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [17]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [18]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [13]:
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [16]:
df['Wteam'].value_counts()

1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: Wteam, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [18]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [17]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [19]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [20]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [23]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [24]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [25]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [26]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [27]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [28]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [65]:
# Write your code here
# By the written instructions:
df1 = df.sort_values('Season').iloc[100:200]['Season']
df2 = df.sort_values('Wteam').iloc[100:200]['Wteam']
df3 = df.sort_values('Wscore').iloc[100:200]['Wscore']
#Since this is an just assignment to show that we can use the right commands, I'll leave it as written in the question, but it is worth noting that sort_values() by default puts the smallest number first, which is usually not what we are interested in when sorting by winning score.

# What I assume was ment:
df1 = df.sort_values('Season').iloc[100:200][['Season', 'Wteam', 'Wscore']]
df2 = df.sort_values('Wteam').iloc[100:200][['Season', 'Wteam', 'Wscore']]
df3 = df.sort_values('Wscore').iloc[100:200][['Season', 'Wteam', 'Wscore']]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

**Note**: Instructions unclear. By the previous instructions, two of the three subsets don't have a winning score column. I have reinterpreted the instructions to keep all three columns in all three subsets.

In [66]:
# Write your code here
from IPython.display import display

display(df1.iloc[[df1['Wscore'].argmax()]]) #Season subset
display(df2.iloc[[df2['Wscore'].argmax()]]) #WTeam subset
display(df3.iloc[[df3['Wscore'].argmax()]]) #WScore subset. Note that the WScore here is a lot smaller since it was literally sorted by smallest first

Season  Wteam  Wscore
2366    1985   1260     105

Season  Wteam  Wscore
26437    1991   1102     103

Season  Wteam  Wscore
90572    2006   1416      45

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [30]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [31]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [62]:
# Write your code here
df['ScoreDifference'] = df['Wscore'] - df['Lscore']
df4 = df[df['ScoreDifference'] > df['ScoreDifference'].mean()]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [73]:
# Write your code here
#df4.sort_values('ScoreDifference', ascending=False).head(1)
df4.iloc[[df4['ScoreDifference'].argmax()]][['Season', 'Wteam', 'Lteam', 'Wscore', 'Lscore', 'ScoreDifference']]

Season  Wteam  Lteam  Wscore  Lscore  ScoreDifference
44653    1996   1409   1341     141      50               91

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [32]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [33]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [34]:
df.values

array([[1985, 20, 1228, ..., 64, 'N', 0],
       [1985, 25, 1106, ..., 70, 'H', 0],
       [1985, 25, 1112, ..., 56, 'H', 0],
       ...,
       [2016, 132, 1246, ..., 77, 'N', 1],
       [2016, 132, 1277, ..., 62, 'N', 0],
       [2016, 132, 1386, ..., 74, 'N', 0]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [35]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [74]:
# Write your code here
df.groupby('Season')['Wscore'].mean().head()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [75]:
# Write your code here
df.groupby('Wteam')['Wscore'].max().head()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
Name: Wscore, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [222]:
# Write your code here
dfmean = df.groupby(['Season', 'Wteam'])['Wscore'].mean()   #Find the mean for each team, per season
dfmax = dfmean.groupby('Season').max()      #Find the max for each season
dfteams = dfmean.groupby('Season').idxmax() #Find the team with the max for each season

pd.DataFrame({      #Join the two dataframes
    'Wteam': [x[1] for x in dfteams], #idxmax() gives tuples for some reason, so we extract the part we want
    'Wscore': dfmax
})


Wteam      Wscore
Season                   
1985     1328   92.800000
1986     1109   91.200000
1987     1380   95.875000
1988     1258  111.750000
1989     1258  117.315789
1990     1258  126.347826
1991     1380  112.312500
1992     1380   99.642857
1993     1380  101.875000
1994     1380  106.583333
1995     1206  102.833333
1996     1206  103.600000
1997     1254   94.142857
1998     1395  101.153846
1999     1317   95.750000
2000     1395   95.647059
2001     1395   99.315789
2002     1242   92.888889
2003     1395   88.222222
2004     1194   92.142857
2005     1311   91.333333
2006     1190   89.533333
2007     1440   99.400000
2008     1440   94.500000
2009     1377   96.200000
2010     1440   95.000000
2011     1449   90.130435
2012     1370   87.400000
2013     1377   86.500000
2014     1322   92.600000
2015     1322   90.733333
2016     1146   97.142857

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [247]:
# Write your code here
df5 = df.groupby(['Season', 'Wteam'])['Wteam'].count().reset_index(name='count')
df5

Season  Wteam  count
0        1985   1102      5
1        1985   1103      9
2        1985   1104     21
3        1985   1106     10
4        1985   1108     19
...       ...    ...    ...
10167    2016   1460     20
10168    2016   1461     12
10169    2016   1462     27
10170    2016   1463     21
10171    2016   1464      9

[10172 rows x 3 columns]

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [255]:
# Write your code here
df5 = df.groupby(['Season', 'Wteam'])['Wteam'].count()
df5idx = df5.groupby('Season').idxmax()

pd.DataFrame({
    'Season': df5idx.index.get_level_values('Season'),
    'Wteam': [x[1] for x in df5idx]
})

Season  Wteam
0     1985   1385
1     1986   1181
2     1987   1424
3     1988   1112
4     1989   1328
5     1990   1247
6     1991   1116
7     1992   1181
8     1993   1231
9     1994   1163
10    1995   1116
11    1996   1269
12    1997   1242
13    1998   1242
14    1999   1181
15    2000   1409
16    2001   1181
17    2002   1153
18    2003   1166
19    2004   1390
20    2005   1228
21    2006   1181
22    2007   1242
23    2008   1272
24    2009   1272
25    2010   1242
26    2011   1242
27    2012   1246
28    2013   1211
29    2014   1455
30    2015   1246
31    2016   1242

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

**Note**: I can't find any questions asking for the average winning score for each team

In [34]:
# Write your code here
dfloss = df.groupby('Lteam')['Lscore'].mean() #Find teams' average loss score
dfwin = df.groupby('Wteam')['Wscore'].mean()  #Find teams' average win score

df6 = pd.DataFrame({  #Join them
    'Wscore': dfwin,
    'Lscore': dfloss,
})
df6['BiggerL'] = df6['Wscore'] < df6['Lscore'] #Add an extra column that shows if a team has a greater average score on losses than wins
df6

Wscore     Lscore  BiggerL
1101  78.111111  60.586207    False
1102  69.893204  59.201507    False
1103  75.839768  64.117347    False
1104  75.825944  64.374317    False
1105  74.960894  61.675373    False
...         ...        ...      ...
1460  75.531469  63.082915    False
1461  75.170082  64.661972    False
1462  79.906021  68.216117    False
1463  71.720102  60.829213    False
1464  73.926056  65.187063    False

[364 rows x 3 columns]

None of the teams has a greater average score on losses than wins. While it is technically possible for a team to score more points on average in games they loose than in games they win, it is very unlikely. It would require them to consistently win when both sides play defensively/passively and loose when both teams play agressively.

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [64]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season    1985
Daynum      20
Wteam     1228
Wscore      81
Lteam     1328
Lscore      64
Wloc         N
Numot        0
Name: 0, dtype: object
Season    1985
Daynum      25
Wteam     1106
Wscore      77
Lteam     1354
Lscore      70
Wloc         H
Numot        0
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [39]:
# Write your code here
for index, row in df.iterrows():
    df['HighScoringGame'][index] = 'Yes' if (row[3] > 100) else 'No'
df

# Alternatively:
#df['HighScoringGame'] = df['Wscore'] > 100
#df

C:\Users\thor\AppData\Local\Temp\ipykernel_7824\4238007000.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['HighScoringGame'][index] = 'Yes' if (row[3] > 100) else 'No'


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

       HighScoringGame  
0                   No  
1                   No  
2                   No  
3                   No  
4                   No  
...                ...  
145284              No  
145285              No  
145286              No  
145287              No  
145288              No  

[145289 rows x 9 columns]

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [74]:
# Write your code here
games_dict = {}

for index, row in df.iterrows():
    games_dict[row[2]] = games_dict.get(row[2], 0) + 1
    games_dict[row[4]] = games_dict.get(row[4], 0) + 1

gameSeries = pd.Series(games_dict, name='GamesPlayed')
gameSeries.index.name = 'Team'
gameSeries.sort_index()

Team
1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: GamesPlayed, Length: 364, dtype: int64

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [156]:
# Write your code here
df_seasons = pd.DataFrame({     #Create new dataframe with the data we want to save
    'Season': df['Season'].unique(),
    'Team': 0,
    'ScoreDiff': 0
})
for index, row in df.iterrows():
    diff = row[3] - row[5]      #Find the score difference
    savedScore = df_seasons.loc[df_seasons['Season'] == row[0], 'ScoreDiff'].values[0] #Retrieve the saved score.
    if diff > savedScore:    #If current diff is higher than saved
        df_seasons.loc[df_seasons['Season'] == row[0], 'Team'] = row[2] #Update the team to current
        df_seasons.loc[df_seasons['Season'] == row[0], 'ScoreDiff'] = diff #Update the score to current

df_seasons

Season  Team  ScoreDiff
0     1985  1361         60
1     1986  1314         84
2     1987  1155         73
3     1988  1328         68
4     1989  1242         70
5     1990  1181         76
6     1991  1163         68
7     1992  1116         82
8     1993  1328         81
9     1994  1228         69
10    1995  1246         74
11    1996  1409         91
12    1997  1278         80
13    1998  1395         66
14    1999  1268         75
15    2000  1261         75
16    2001  1196         75
17    2002  1328         59
18    2003  1403         72
19    2004  1328         70
20    2005  1403         64
21    2006  1400         72
22    2007  1401         74
23    2008  1182         70
24    2009  1314         68
25    2010  1397         75
26    2011  1388         69
27    2012  1345         62
28    2013  1235         58
29    2014  1256         74
30    2015  1181         69
31    2016  1139         73

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [157]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [169]:
# Write your code here
s_games = df['Wteam'].value_counts().sort_index() + df['Lteam'].value_counts().sort_index()
s_games.name = 'GamesPlayed'
s_games.index.name = 'Team'
s_games

Team
1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: GamesPlayed, Length: 364, dtype: int64

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [178]:
# Write your code here
df['ScoreDiff'] = df['Wscore'] - df['Lscore']   #Find score diff for each game
s_max = df.groupby('Season')['ScoreDiff'].idxmax()  #Find index for game with highest score diff
df.loc[s_max]   #List found games

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410      87    H      0   
57618     1999      58   1268     132   1317      57    H      0   
60616     2000      19   1261     112   1212      37    H      0   
65999     2001      41   1196     125   1197      50    H      0   
70737     2002      50   1328     109   1183      50    H      0   
74791     2003      33   1403     107   1311      35    H      0   
79181     2004      29   1328      94   1115      24    H      0   
83372     2005      18   1403     119   1421      55    H      0   
89542     2006      60   1400     110   1341      38    H      0   
94507     2007      59   1401     101   1212      27    H      0   
97737     2008       4   1182     129   1224      59    H      0   
103723    2009      27   1314     116   1421      48    H      0   
108405    2010      15   1397     124   1421      49    H      0   
114015    2011      22   1388     121   1152      52    H      0   
118699    2012      11   1345      96   1296      34    H      0   
124057    2013       7   1235      98   1105      40    H      0   
131373    2014      61   1256     126   1255      52    H      0   
134590    2015      11   1181     113   1342      44    H      0   
140040    2016      12   1139     144   1154      71    H      0   

       HighScoringGame  ScoreDiff  
236                Yes         60  
4731               Yes         84  
8240               Yes         73  
12046              Yes         68  
16677              Yes         70  
19502              Yes         76  
25161              Yes         68  
27997              Yes         82  
33858              Yes         81  
36404              Yes         69  
39858              Yes         74  
44653              Yes         91  
49033              Yes         80  
52600              Yes         66  
57618              Yes         75  
60616              Yes         75  
65999              Yes         75  
70737              Yes         59  
74791              Yes         72  
79181               No         70  
83372              Yes         64  
89542              Yes         72  
94507              Yes         74  
97737              Yes         70  
103723             Yes         68  
108405             Yes         75  
114015             Yes         69  
118699              No         62  
124057              No         58  
131373             Yes         74  
134590             Yes         69  
140040             Yes         73

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [38]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [39]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [40]:
type(df['Wscore'])

pandas.core.series.Series

In [41]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [42]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

Here's an equivalent using iloc

In [43]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [44]:
df.isnull().sum()

Season    0
Daynum    0
Wteam     0
Wscore    0
Lteam     0
Lscore    0
Wloc      0
Numot     0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y